api/hasura/cron/syncCoSouls.ts

```diff
@@ -0,0 +1,233 @@
+    } else {
+      errors.push(cosoul.id);
+    }
+  }
+  const message = `${cosouls.length} CoSouls updated`;
+  return { message, updated, errors };
+}
+
+export async function isHistoricalPGiveFinished() {
+  const previousMonth = DateTime.local().minus({ months: 1 });
+  const startFrom = previousMonth.startOf('month');
+  const endTo = previousMonth.endOf('month');
+
+  /* Filter for circles that are still pending PGIVE calculation for last month */
+  const lastMonthCircleIds = await getCirclesNoPgiveWithDateFilter(
+    startFrom,
+    endTo
+  );
+
+  // If no circles are pending PGIVE calculation, we can proceed
+  return lastMonthCircleIds.length === 0;
+}
+
+const getCoSoulsToUpdate = async () => {
+  const previousMonth = DateTime.local().minus({ months: 1 });
+  const endOfPreviousMonth = previousMonth.endOf('month');
+
+  const { cosouls } = await adminClient.query(
+    {
+      cosouls: [
+        {
+          limit: LIMIT_USERS_TO_SYNC,
+          where: {
+            _or: [
+              {
+                checked_at: {
+                  _is_null: true,
+                },
+              },
+              {
+                checked_at: {
+                  _lt: endOfPreviousMonth.toISO(),
+                },
+              },
+            ],
+          },
+        },
+        {
+          id: true,
+          token_id: true,
+          profile_id: true,
+          pgive: true,
+          profile: {
+            address: true,
+          },
+        },
+      ],
+    },
+    {
+      operationName: 'cron__syncCoSouls__getCoSoulsToUpdate',
+    }
+  );
+  return cosouls;
+};
+
+const syncCoSoulToken = async (
+  coSoulId: number,
+  address: string,
+  totalPGIVE: number,
+  tokenId: number
+) => {
+  if (totalPGIVE > 0) {
+    totalPGIVE = Math.floor(totalPGIVE);
+    await setOnChainPGIVE(tokenId, totalPGIVE);
+    console.log(
+      'set PGIVE on chain for tokenId: ' +
+        tokenId +
+        ' address: ' +
+        address +
+        ' to ' +
+        totalPGIVE
+    );
+  } else {
+    console.log(
+      'skipping setting on-chain PGIVE because it is 0 for tokenId: ' +
+        tokenId +
+        ' address: ' +
+        address
+    );
+  }
+
+  await adminClient.mutate(
+    {
+      update_cosouls_by_pk: [
+        {
+          pk_columns: {
+            id: coSoulId,
+          },
+          _set: {
+            pgive: totalPGIVE,
+            checked_at: new Date().toISOString(),
+            synced_at: new Date().toISOString(),
+          },
+        },
+        {
+          id: true,
+        },
+      ],
+    },
+    {
+      operationName: 'cron__updateCoSoulCache',
+    }
+  );
+  console.log(
+    'Updated CoSoul with address: ' + address + ' to tokenId: ' + tokenId
+  );
+};
+
+async function updateCoSoulOnChain(cosoul: CoSoul, totalPGIVE: number) {
+  try {
+    await syncCoSoulToken(
+      cosoul.id,
+      cosoul.profile.address,
+      totalPGIVE,
+      cosoul.token_id
+    );
+    return true;
```

Comment:
Maybe we ping Opensea here and tell them to refetch metadata?

https://api.opensea.io/api/v1/asset/<your_contract_address>/<token_id>/?force_update=true

https://docs.opensea.io/docs/metadata-standards#metadata-updates

Or we emit an event from the contract.